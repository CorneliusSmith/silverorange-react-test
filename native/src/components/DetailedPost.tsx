import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

export function DetailedPost() {
  return (
    <View style={styles.container}>
      <Text>DetailedPost</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
