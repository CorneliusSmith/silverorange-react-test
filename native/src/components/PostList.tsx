import React, { Component } from 'react';
import { ActivityIndicator, StyleSheet, Text, View, FlatList, TouchableOpacity,} from 'react-native';
import { ProfileScreenNavigationProp, Props, State,} from '../types/PostListTypes';

export const sortReverseChronologically = (json: any) => {
  json.sort(
    (later: any, older: any) =>
      new Date(older.publishedAt).getTime() -
      new Date(later.publishedAt).getTime()
  );
  return json
}

export class PostList extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      data: [],
      isLoaded: false,
      isFiltered: false,
      authorName: '',
    };
  }

  componentDidMount() {
    fetch('http://10.0.0.215:4000/posts')
      .then((response) => response.json())  
      .then((json) => {
        json = sortReverseChronologically(json)
        this.setState({
          isLoaded: true,
          data: json,
        });
      });
  }

  filterByAuthor(name: string, posts: any[]) {
    posts = posts.filter((post) => post.author.name == name);
    return (
      <FlatList
        data = {posts}
        renderItem={({ item }) => (
            <View style={styles.item}>
              <Text style={{ fontWeight: 'bold' }}>
                Author: {item.author.name}
              </Text>
              <Text> Title: {item.title} </Text>
              <Text>
                Date Published: {new Date(item.publishedAt).toDateString()}
              </Text>
              <Text> Summary: {item.body.split('\n', 1)} </Text>
            </View>
        )}
      />
    );
  }


  render() {
    var { isLoaded, data, isFiltered, authorName } = this.state;

    if (!isLoaded) {
      return <ActivityIndicator />;
    }

    if (isFiltered) {
      return (
        <View style={styles.container}>
          <TouchableOpacity
            onPress={() =>
              this.setState({
                isFiltered: false,
                authorName: '',
              })
            }
          >
          <Text style={{ fontWeight: 'bold' }}>
            Tap Here To Return To Unfiltered List
          </Text>
        </TouchableOpacity>
        {this.filterByAuthor(authorName, data)}
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <Text style={{ fontWeight: 'bold' }}>
          Tap An Authors Name To View Only Their Posts
        </Text>
        <FlatList
          data={this.state.data}
          renderItem={({ item }) => (
              <View style={styles.item}>
                <TouchableOpacity
                  onPress={() =>
                    this.setState({
                      isFiltered: true,
                      authorName: item.author.name,
                    })
                  }
                >
                  <Text style={{ fontWeight: 'bold' }}>
                    Author: {item.author.name}
                  </Text>
                </TouchableOpacity>
                <Text> Title: {item.title} </Text>
                <Text>
                  Date Published: {new Date(item.publishedAt).toDateString()}
                </Text>
                <Text> Summary: {item.body.split('\n', 1)} </Text>
              </View>
          )}
        />
      </View>
    );
  }
}
export default PostList;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },

  item: {
    marginTop: 30,
    padding: 30,
    backgroundColor: 'silver',
    borderRadius: 10,
    borderWidth: 3,
    borderColor: 'orange',
  },
});
