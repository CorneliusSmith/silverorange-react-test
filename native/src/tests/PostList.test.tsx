import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Platform } from 'react-native';
import { PostList, sortReverseChronologically} from '../components/PostList';
import 'isomorphic-fetch';
import 'react-native-gesture-handler/jestSetup';

jest.useFakeTimers()
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');


describe('<PostList/>', () => {
  const tree = renderer.create(<PostList/>).toJSON();
  test('if it renders without crashing', async () => {
    await act(async ()=>{
      expect(tree).not.toBeNull();
    });
  });

  test('if it matches the snapshot', async () => {
    await act(async ()=>{
      expect(tree).toMatchSnapshot();
    });
  });

});


describe('sortReverseChronologically', () => {
  const postDataOg = [
    {
      title: 'post1',
      body: 'yello',
      author: {
        name: 'numbah1',
        id: '1',
      },
      id: 'a',
      publishedAt: new Date('08-28-2020'),
    },
    {
      title: 'post2',
      body: 'hello',
      author: {
        name: 'numbah2',
        id: '2',
      },
      id: 'b',
      publishedAt: new Date('08-29-2020'),
    },
    {
      title: 'post3',
      body: 'chello',
      author: {
        name: 'numbah3',
        id: '3',
      },
      id: 'c',
      publishedAt: new Date('08-30-2020'),
    },
  ];

  const postDataReverseChronological = [
    {
      title: 'post3',
      body: 'chello',
      author: {
        name: 'numbah3',
        id: '3',
      },
      id: 'c',
      publishedAt: new Date('08-30-2020'),
    },
    {
      title: 'post2',
      body: 'hello',
      author: {
        name: 'numbah2',
        id: '2',
      },
      id: 'b',
      publishedAt: new Date('08-29-2020'),
    },
    {
      title: 'post1',
      body: 'yello',
      author: {
        name: 'numbah1',
        id: '1',
      },
      id: 'a',
      publishedAt: new Date('08-28-2020'),
    },
  ];
  
  test('if it sorts correctly',  () => {
    expect(sortReverseChronologically(postDataOg)).toStrictEqual(postDataReverseChronological)
  });

});