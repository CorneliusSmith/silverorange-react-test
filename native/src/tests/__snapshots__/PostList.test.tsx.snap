// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostList/> if it matches the snapshot 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="small"
/>
`;
